### DISCUSSION TITLE
Firebase Emulators UI being blocked in WSL 2

### DISCUSSION THREAD

**Discussion by b0ot** [microsoft/WSL#6240] [General]
# Environment

```
Windows build number: 19042.630
Your Distribution version: Ubuntu 20.04
Whether the issue is on WSL 2 and/or WSL 1: WSL 2
```

# Steps to reproduce
I followed the [Emulators Code Lab](https://firebase.google.com/learn/codelabs/firebase-emulator#1) tutorial with a clean pull from git into a new folder. I follow the steps to get the Emulator UI running. There does not appear to be any obvious errors except that after starting the Emulators UI, I'm still unable to access any of the associated webpages e.g.
│ ✔  All emulators ready! View status and logs at http://127.0.0.1:4000
├───────────┼────────────────┼─────────────────────────────────┤
│ Functions │ 127.0.0.1:5001 │ http://127.0.0.1:4000/functions │
├───────────┼────────────────┼─────────────────────────────────┤
│ Firestore │ 127.0.0.1:8080 │ http://127.0.0.1:4000/firestore │
├───────────┼────────────────┼─────────────────────────────────┤

ADDITIONAL INFORMATION

- After initial failure to bring up UI, I tried again to have it work after running the networking Script here
That leads to running a custom script here: [#4150 (comment) ](https://github.com/microsoft/WSL/issues/4150#issuecomment-667621988)
After adding additional ports: 22,80,443,8080,10000,3000,4000,5000,5001,8081,8082,4200,3128,4400,4500 

However, this also did not seem to work.

I restarted the computer and then verified the WSL install by re-running steps [described here](https://docs.microsoft.com/en-us/windows/wsl/install-win10)

I then installed a completely new Debian distro for WSL2
I installed
- curl, nvm, node lts, java, git then followed the Emulators Codelab again

I was still unable to access the Emulator UI

I tried multiple browsers (Chrome / Firefox) and multiple variations of 127.0.0.1 / localhost all without luck.

**WSL logs**: 

#  Expected behavior

Should be able to see the Emulator UI via the Web Browswer
# Actual behavior

Unable to access. 

Note: I have been debugging with Google's Firebase team for a few weeks and they suggested I post here. @yuchenshi @samtstern
[Last Thread](https://github.com/firebase/firebase-tools/issues/2793)

Previous Threads:
[Emulator UI does not show up on Windows w/ WSL2 (current 332 workarounds do not work) #377](https://github.com/firebase/firebase-tools-ui/issues/377)
[firestore emulator on wsl2 does not work, results in ERR_CONNECTION_REFUSED #332](https://github.com/firebase/firebase-tools-ui/issues/332)

Debug
> Could you please try these debugging steps?
> 
> 1. `curl http://127.0.0.1:4000` and `netstat -nl` within WSL2 (e.g. using a bash session on Linux)
> 2. `curl http://127.0.0.1:4000` from the host (e.g. PowerShell on Windows)
> 
> Please paste the output.
> 
> And by the way, how are you accessing the Emulator UI? Are you accessing it on the Windows host on the same device where WSL2 is running? Either way, make sure you use `http://127.0.0.1/` instead of `localhost`.

 **1. `curl http://127.0.0.1:4000` and `netstat -nl` within WSL2 (e.g. using a bash session on Linux)**
```
tom-linux@DESKTOP:~$ curl http://127.0.0.1:4000
<!doctype html><html lang="en"><head><meta charset="utf-8"/><meta name="msapplication-TileColor" content="#681da8"><meta name="viewport" content="width=device-width,initial-scale=1"/><meta name="theme-color" content="#681da8"/><meta name="description" content="Web site created using create-react-app"/><link rel="manifest" href="/manifest.json?v=3"/><link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png?v=3"><link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png?v=3"><link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png?v=3"><link rel="mask-icon" href="/safari-pinned-tab.svg?v=3" color="#681da8"><link rel="icon" href="/favicon.ico?v=3"/><title>Firebase Emulator Suite</title><link href="https://fonts.googleapis.com/css?family=Roboto:300,400,500" rel="stylesheet"/><link href="https://fonts.googleapis.com/css?family=Roboto Mono:300,400,500" rel="stylesheet"><link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet"><link href="/static/css/2.3a9d61e3.chunk.css" rel="stylesheet"><link href="/static/css/main.8b1c7704.chunk.css" rel="stylesheet"></head><body><noscript>You need to enable JavaScript to run this app.</noscript><div id="root"></div><script>!function(e){function r(r){for(var n,i,a=r[0],f=r[1],c=r[2],s=0,p=[];s<a.length;s++)i=a[s],Object.prototype.hasOwnProperty.call(o,i)&&o[i]&&p.push(o[i][0]),o[i]=0;for(n in f)Object.prototype.hasOwnProperty.call(f,n)&&(e[n]=f[n]);for(l&&l(r);p.length;)p.shift()();return u.push.apply(u,c||[]),t()}function t(){for(var e,r=0;r<u.length;r++){for(var t=u[r],n=!0,a=1;a<t.length;a++){var f=t[a];0!==o[f]&&(n=!1)}n&&(u.splice(r--,1),e=i(i.s=t[0]))}return e}var n={},o={1:0},u=[];function i(r){if(n[r])return n[r].exports;var t=n[r]={i:r,l:!1,exports:{}};return e[r].call(t.exports,t,t.exports,i),t.l=!0,t.exports}i.e=function(e){var r=[],t=o[e];if(0!==t)if(t)r.push(t[2]);else{var n=new Promise((function(r,n){t=o[e]=[r,n]}));r.push(t[2]=n);var u,a=document.createElement("script");a.charset="utf-8",a.timeout=120,i.nc&&a.setAttribute("nonce",i.nc),a.src=function(e){return i.p+"static/js/"+({}[e]||e)+"."+{3:"f1fb2c8e",4:"ff183786",5:"3c3415d4",6:"1f361fdf",7:"89d4cf82",8:"dfaa6f1d",9:"1bec8998",10:"947d73d2"}[e]+".chunk.js"}(e);var f=new Error;u=function(r){a.onerror=a.onload=null,clearTimeout(c);var t=o[e];if(0!==t){if(t){var n=r&&("load"===r.type?"missing":r.type),u=r&&r.target&&r.target.src;f.message="Loading chunk "+e+" failed.\n("+n+": "+u+")",f.name="ChunkLoadError",f.type=n,f.request=u,t[1](f)}o[e]=void 0}};var c=setTimeout((function(){u({type:"timeout",target:a})}),12e4);a.onerror=a.onload=u,document.head.appendChild(a)}return Promise.all(r)},i.m=e,i.c=n,i.d=function(e,r,t){i.o(e,r)||Object.defineProperty(e,r,{enumerable:!0,get:t})},i.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},i.t=function(e,r){if(1&r&&(e=i(e)),8&r)return e;if(4&r&&"object"==typeof e&&e&&e.__esModule)return e;var t=Object.create(null);if(i.r(t),Object.defineProperty(t,"default",{enumerable:!0,value:e}),2&r&&"string"!=typeof e)for(var n in e)i.d(t,n,function(r){return e[r]}.bind(null,n));return t},i.n=function(e){var r=e&&e.__esModule?function(){return e.default}:function(){return e};return i.d(r,"a",r),r},i.o=function(e,r){return Object.prototype.hasOwnProperty.call(e,r)},i.p="/",i.oe=function(e){throw console.error(e),e};var a=this["webpackJsonpfirebase-tools-ui"]=this["webpackJsonpfirebase-tools-ui"]||[],f=a.push.bind(a);a.push=r,a=a.slice();for(var c=0;c<a.length;c++)r(a[c]);var l=f;t()}([])</script><script src="/static/js/2.fbaea028.chunk.js">
```

tom-linux@DESKTOP:~$ netstat -nl
```
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State
tcp        0      0 127.0.0.1:5000          0.0.0.0:*               LISTEN
tcp        0      0 127.0.0.1:5001          0.0.0.0:*               LISTEN
tcp        0      0 127.0.0.1:8080          0.0.0.0:*               LISTEN
tcp        0      0 127.0.0.1:4400          0.0.0.0:*               LISTEN
tcp        0      0 127.0.0.1:4500          0.0.0.0:*               LISTEN
tcp        0      0 127.0.0.1:39135         0.0.0.0:*               LISTEN
tcp        0      0 127.0.0.1:41119         0.0.0.0:*               LISTEN
tcp        0      0 127.0.0.1:4000          0.0.0.0:*               LISTEN
Active UNIX domain sockets (only servers)
Proto RefCnt Flags       Type       State         I-Node   Path
unix  2      [ ACC ]     SEQPACKET  LISTENING     18572    /run/WSL/779_interop
unix  2      [ ACC ]     SEQPACKET  LISTENING     18653    /run/WSL/7_interop
```

2. curl http://127.0.0.1:4000` from the host (e.g. PowerShell on Windows)
```
PS C:\Users\tom> curl http://127.0.0.1:4000
curl : The underlying connection was closed: An unexpected error occurred on a receive.
At line:1 char:1
+ curl http://127.0.0.1:4000
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-WebRequest], WebExc
   eption
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand`
```

**Comment by therealkenc** [comment]
> 19042.630

Appreciate the netstat output, that helps a lot. Given the build above, it is likely you'll need to have the service listening on `0.0.0.0`. Yours are listening on `127.0.0.1`. This is configurable via firebase, although you'll get better eyeballs on firebase configuration in a firebase venue. This is #5298 (not open #4150), fixed circa 20180, which presumably did not make 19042. 

General off-the-record guidance for 19042 (ref #5706, [message](https://github.com/microsoft/WSL/issues/5706#issuecomment-668353533)) would be to eschew `localhost`. A better way to roll is use [`avahi-daemon`](http://manpages.ubuntu.com/manpages/bionic/man8/avahi-daemon.8.html) spiritually similar the `sandstorm` thread #5368 ([message](https://github.com/microsoft/WSL/issues/5368#issuecomment-645316048)). That will get you real IP address instead of `localhost`, or typing the server address manually. TL;DR, `sudo service dbus start` and `sudo service avahi-daemon start`, after editing `/etc/avahi/avahi-daemon.conf`.

I almost commented over in #5706 that running #4150 related scripts doesn't make a lot of sense in the local machine context. YRMV. Folks advocating that are welcome weigh in on the discussion if they like.
